addons/stock_picking_wave/models/stock_picking_wave.py

```diff
@@ -1,77 +1,66 @@
-        self.pool.get('stock.picking').action_cancel(cr, uid, picking_todo, context=context)
-        return self.write(cr, uid, ids, {'state': 'cancel'}, context=context)
+    @api.multi
+    def cancel_picking(self):
+        self.mapped('picking_ids').action_cancel()
+        return self.write({'state': 'cancel'})
 
-    def print_picking(self, cr, uid, ids, context=None):
+    @api.multi
+    def print_picking(self):
         '''
         This function print the report for all picking_ids associated to the picking wave
         '''
-        context = dict(context or {})
-        picking_ids = []
-        for wave in self.browse(cr, uid, ids, context=context):
-            picking_ids += [picking.id for picking in wave.picking_ids]
-        if not picking_ids:
+        pickings = self.mapped('picking_ids')
+        if not pickings:
             raise UserError(_('Nothing to print.'))
-        context['active_ids'] = picking_ids
-        context['active_model'] = 'stock.picking'
-        return self.pool.get("report").get_action(cr, uid, [], 'stock.report_picking', context=context)
+        return self.env["report"].with_context(active_ids=pickings.ids, active_model='stock.picking').get_action(pickings, 'stock.report_picking')
 
-    def create(self, cr, uid, vals, context=None):
+    @api.model
+    def create(self, vals):
         if vals.get('name', '/') == '/':
-            vals['name'] = self.pool.get('ir.sequence').next_by_code(cr, uid, 'picking.wave') or '/'
-        return super(stock_picking_wave, self).create(cr, uid, vals, context=context)
+            vals['name'] = self.env['ir.sequence'].next_by_code('picking.wave') or '/'
+        return super(StockPickingWave, self).create(vals)
 
-    def done(self, cr, uid, ids, context=None):
-        picking_todo = set()
-        for wave in self.browse(cr, uid, ids, context=context):
-            for picking in wave.picking_ids:
-                if picking.state in ('cancel', 'done'):
-                    continue
-                if picking.state != 'assigned':
-                    raise UserError(_('Some pickings are still waiting for goods. Please check or force their availability before setting this wave to done.'))
-                message_body = "<b>%s:</b> %s <a href=#id=%s&view_type=form&model=stock.picking.wave>%s</a>" % (_("Transferred by"), _("Picking Wave"), wave.id, wave.name)
-                picking.message_post(body=message_body)
-                picking_todo.add(picking.id)
-        if picking_todo:
-            self.pool.get('stock.picking').action_done(cr, uid, list(picking_todo), context=context)
-        return self.write(cr, uid, ids, {'state': 'done'}, context=context)
+    @api.multi
+    def done(self):
+        pickings_todo = self.env['stock.picking']
+        for picking in self.mapped('picking_ids').filtered(lambda x: x.state not in ('cancel', 'done')):
+            if picking.state != 'assigned':
```

Comment:
I would move the check outside of the main operational loop, using an if (any (picking.state ... )).